### PR TITLE
[#120] Implement IoC Container

### PR DESCRIPTION
**Pull request by hwbrzzl**
Close: https://github.com/goravel/goravel/issues/120

### CODE REVIEW

File: facades/artisan.go [hunk 1]
```diff
@@ -5,3 +5,7 @@ import (
 )
 
 var Artisan console.Artisan
+
+func NewArtisan() console.Artisan {
+	return App().MakeArtisan()
+}
```
Comment (hwbrzzl) [inline facades/artisan.go:11]: Temporarily add `NewArtisan`, will replace `Artisan`: `func NewArtisan() console.Artisan`

File: facades/config.go [hunk 2]
```diff
@@ -1,5 +1,11 @@
...
+import (
+	configcontract "github.com/goravel/framework/contracts/config"
+)
 
-var Config config.Config
+var Config = NewConfig()
+
+func NewConfig() configcontract.Config {
+	return App().MakeConfig()
+}
```
Comment (hwbrzzl) [inline facades/config.go:11]: Temporarily add `NewArtisan`, will replace `Config`: `func Config() configcontract.Config`

**Review by DevHaoZi — COMMENTED** [review]
Looks no problems.

**Review by DevHaoZi — APPROVED** [review]

File: auth/auth.go [hunk 3]
```diff
@@ -10,8 +10,10 @@ import (
 	"gorm.io/gorm/clause"
 
 	contractsauth "github.com/goravel/framework/contracts/auth"
+	"github.com/goravel/framework/contracts/cache"
+	"github.com/goravel/framework/contracts/config"
+	"github.com/goravel/framework/contracts/database/orm"
 	"github.com/goravel/framework/contracts/http"
-	"github.com/goravel/framework/facades"
```
Comment (hwbrzzl) [inline auth/auth.go:14]: All of the modules shouldn't be dependent on facades and should inject if needed.

File: auth/auth_test.go [hunk 4]
```diff
@@ -27,59 +28,64 @@ type User struct {
...
-
+	s.mockCache = &cachemock.Cache{}
+	s.mockConfig = &configmock.Config{}
+	s.mockOrm = &ormmock.Orm{}
+	s.mockDB = &ormmock.Query{}
+	s.auth = NewAuth(guard, s.mockCache, s.mockConfig, s.mockOrm)
 }
 
 func (s *AuthTestSuite) TestLoginUsingID_EmptySecret() {
-	mockConfig := mock.Config()
```
Comment (hwbrzzl) [inline auth/auth_test.go:45]: All of the test cases shouldn't be dependent on the `github.com/goravel/framework/testing/mock` package, will cause import cycle error, use `&cachemock.Cache{}` instead.

File: auth/service_provider.go [hunk 5]
```diff
@@ -6,24 +6,33 @@ import (
...
-	facades.Gate = access.NewGate(context.Background())
+func (database *ServiceProvider) Register(app foundation.Application) {
+	app.Singleton(BindingAuth, func() (any, error) {
+		config := app.MakeConfig()
+		return NewAuth(config.GetString("auth.defaults.guard"),
+			app.MakeCache(), config, app.MakeOrm()), nil
+	})
+	app.Singleton(BindingGate, func() (any, error) {
+		return access.NewGate(context.Background()), nil
+	})
```
Comment (hwbrzzl) [inline auth/service_provider.go:26]: Bind and inject like this.

File: database/gorm/wire.go
Comment (hwbrzzl) [inline database/gorm/wire.go:1]: Use wire to implement dependency injection.

**Review by DevHaoZi — APPROVED** [review]
LGTM

**Review by DevHaoZi — APPROVED** [review]
LGTM

**Pull request merged**